src/volume_calc.cpp

```diff
@@ -0,0 +1,423 @@
+#include "openmc/random_lcg.h"
+#include "openmc/settings.h"
+#include "openmc/timer.h"
+#include "openmc/xml_interface.h"
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+#include "xtensor/xadapt.hpp"
+#include "xtensor/xview.hpp"
+
+#include <algorithm> // for copy
+#include <cmath> // for pow, sqrt
+#include <sstream>
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+namespace model {
+std::vector<VolumeCalculation> volume_calcs;
+}
+
+//==============================================================================
+// VolumeCalculation implementation
+//==============================================================================
+
+VolumeCalculation::VolumeCalculation(pugi::xml_node node) {
+  // Read domain type (cell, material or universe)
+  std::string domain_type = get_node_value(node, "domain_type");
+  if (domain_type == "cell") {
+    domain_type_ = FILTER_CELL;
+  }
+  else if (domain_type == "material") {
+    domain_type_ = FILTER_MATERIAL;
+  }
+  else if (domain_type == "universe") {
+    domain_type_ = FILTER_UNIVERSE;
+  }
+  else {
+    fatal_error(std::string("Unrecognized domain type for stochastic "
+      "volume calculation:" + domain_type));
+  }
+
+  // Read domain IDs, bounding corodinates and number of samples
+  domain_ids_ = get_node_array<int>(node, "domain_ids");
+  lower_left_ = get_node_array<double>(node, "lower_left");
+  upper_right_ = get_node_array<double>(node, "upper_right");
+  n_samples_ = std::stoi(get_node_value(node, "samples"));
+}
+
+void VolumeCalculation::check_hit(int i_material, std::vector<int>& indices,
+  std::vector<int>& hits) const {
+
+  // Check if this material was previously hit and if so, increment count
+  bool already_hit = false;
+  for (int j = 0; j < indices.size(); j++) {
+    if (indices[j] == i_material) {
+      hits[j]++;
+      already_hit = true;
+    }
+  }
+
+  // If the material was not previously hit, append an entry to the material
+  // indices and hits lists
+  if (!already_hit) {
+    indices.push_back(i_material);
+    hits.push_back(1);
+  }
+}
+
+std::vector<VolumeCalculation::Result> VolumeCalculation::execute() const
+{
+  // Shared data that is collected from all threads
+  int n = domain_ids_.size();
+  std::vector<std::vector<int>> master_indices(n); // List of material indices for each domain
+  std::vector<std::vector<int>> master_hits(n); // Number of hits for each material in each domain
+
+  // Divide work over MPI processes
+  int min_samples = n_samples_ / mpi::n_procs;
+  int remainder = n_samples_ % mpi::n_procs;
+  int i_start, i_end;
+  if (mpi::rank < remainder) {
+    i_start = (min_samples + 1)*mpi::rank;
+    i_end = i_start + min_samples + 1;
+  }
+  else {
+    i_start = (min_samples + 1)*remainder + (mpi::rank - remainder)*min_samples;
+    i_end = i_start + min_samples;
+  }
```

Comment:
Double check the math in here.  I tried running a test with 101 samples and 2 procs and I get that `i_end = 101` for the second proc which means only 100 samples will be run.

Reply:
I think it's right actually. In that case, proc 1 has samples [0, 51) (51 total) and proc 2 has samples [51, 101) (50 total).

Reply:
Ah whoops, you're right!  This logic looks good then